texlive: make packages and schemes work as build inputs

xworld21 opened the pull request:
## Description of changes
Add appropriate `dev` outputs and setup hooks so that (most) TeX Live packages can be safely used as build inputs without combining them. This is a made possible by #250963. Draft because it is stacked on previous PRs, it causes a rebuild, and it requires at least `texlive.__pkgs` to stabilise – better yet if we wait until we have determined the new top level interface of `texlive`.

Example use case:

```nix
nativeBuildInputs = [ texlive.combined.scheme-small ] ++ (with texlive.__pkgs; [ libertinus-fonts a2ping ]);
```

Here safely means that if a package or one of its transitive dependencies causes a rebuild of the formats, then it causes an evaluation error when used as build input; hence
```nix
nativeBuildInputs = [ texlive.combined.scheme-small ] ++ (with texlive.__pkgs; [ collection-langitalian.dev ]);
# error: The TeX Live package 'collection-langitalian' and/or its dependencies cannot be used as a build input. Please add 'collection-langitalian' to texlive.combine instead.
```
Similarly, we stop evaluation for the format themselves:
```nix
nativeBuildInputs = [ texlive.combined.scheme-small ] ++ (with texlive.__pkgs; [ pdftex ]);
# error: The TeX Live package 'pdftex' and/or its dependencies cannot be used as a build input. Please add 'pdftex' to texlive.combine instead.
```
and for whichever packages we manually flagged as not `allowedAsBuildInput` (see `chktex`).

Font packages seem to work with (lua)latex, but not with xelatex of course, because of fontconfig.

#### TODOs
- documentation!
- a utility like `makeTeXHook`, for adding the setup hooks to non-TeX Live packages
- update `tests.texlive.binaries` to discover which binaries break outside of schemes (e.g. `chktex` fails badly)
- of course, this is all pending merge & stabilisation of `texlive.pkgs`

## Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/combine.nix:
```diff
@@ -145,6 +145,7 @@ in (buildEnv {
     # patch texmf-dist  -> $TEXMFDIST
     # patch texmf-local -> $out/share/texmf-local
     # patch texmf.cnf   -> $TEXMFSYSVAR/web2c/texmf.cnf
+    # add TEXMFAUXTREES to texmfcnf.lua to support setup hooks
```
Have you tested whether that works? If I remember correctly, ConTeXt only uses its file database for lookups and doesn't do filesystem searching.

xworld21 replied:
Right: `TEXMFAUXTREES=stuff mtxrun --find-file stuff` confirms that the variable works as intended, provided it's added to `texmfcnf.lua` as in the commit, so we are good with ConTeXt.

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/build-texlive-package.nix:
```diff
@@ -202,5 +219,51 @@ let
+  # add tex folders to TEXMFAUXTREES, OSFONTDIR (for luaotfload)
+  dev = let container =
+    if disallowedAsBuildInput then
+      builtins.throw "The TeX Live package '${pname}' and/or its dependencies cannot be used as a build input. Please add '${pname}' to texlive.combine instead."
```
There might be valid reasons to use one of the disallowed packages as a build input, for example if one wants to use a tex engine (without a preloaded format), so failing eval for this might be a little harsh? 

On the other side, such valid uses should be extremele rare, and the restriction could still be removed once such a use case emerges.

xworld21 replied:
True, one can compile formats on the side, like in the HOME folder. It's a dangerous use case (once you compile formats on your own, things get confusing) but we should make that configurable.

xworld21 replied:
> for example if one wants to use a tex engine (without a preloaded format), so failing eval for this might be a little harsh?

Expanding on this: if you launch `pdftex` from `texlive.bin.core`, it will try calling `mktexfmt` in user mode (i.e. output in `~/.texlive2022/...`), which then reads `fmtutil.cnf` and fails because `pdftex` is disabled. This is without getting into the `texlinks` side of the equation.

The solution would be an environment hook to `texlive.combine` that inspects the other build inputs for engines, generates a custom `fmtutil.cnf`, and runs `texlinks` on a temporary folder added in PATH... too much work I suspect!

PS: actually, one can override the error by specifying the output `texlive.__pkgs.FORMAT.out`.

xworld21 replied:
A related idea: using TeX Live as a build input in a shell has the side effect of adding potential garbage to `~/.texlive2022`. Even looking up a font means it will be added permanently to the database. How about setting `TEXMFVAR` to a temporary folder? It doesn't change anything for builds, it protects `HOME` when running a shell. The cost is the caches being lost when rerunning `nix-shell` (unless there is some deterministic temp folder we can use).

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/build-texlive-package.nix:
```diff
@@ -202,5 +219,51 @@ let
+  # exceptions are:
+  # - format binaries, which must be provided by a combined build input
+  # - hyphenation patterns, which cause formats to be rebuilt
+  # - occasional packages that look up in TEXMFROOT or similar
```
Are there any fundamental problems with just adding format containers that are built with all the hyphenation patterns (and leave an escape hatch that drops the containers and runs fmtutil if someone wants to add their own custom hyphenation patterns)?

apfelkuchen6 replied:
arrgh, this wouldn't allow the paper size customization

xworld21 replied:
> format containers that are built with all the hyphenation patterns

That could be an option, just force all hyphenation patterns in every scheme. If the size is small I don't see why not, surely adding more patterns has no other side effect.

I would still build the formats by calling `buildEnv` recursively, that way we are sure that they use the same configuration, including paper size.

apfelkuchen6 reviewed (commented):
That's a neat trick to allow using "minimal" tex dev environments without spending years regenerating formats. Nice.

apfelkuchen6 commented on pkgs/tools/typesetting/tex/texlive/package-setup-hook.sh:
```diff
@@ -0,0 +1,12 @@
+  for texFolder in "${texFolders[@]}" ; do
+    addToSearchPath TEXMFAUXTREES "$texFolder"
+    if [[ -d "$texFolder/fonts" ]] ; then
+      addToSearchPath OSFONTDIR "$texFolder/fonts"
```
OSFONTDIR is interpolated in OPENTYPEFONTS in texmf.cnf as
```
% OpenType outline fonts.
OPENTYPEFONTS = $TEXMFDOTDIR;$TEXMF/fonts/{opentype,truetype}//;$OSFONTDIR//
```
Thus if if OSFONTDIR contains multiple directories, they must be separated by `//:` as otherwise recursion is only enabled for the last one.

apfelkuchen6 replied:
Actually, I think that all of this can be removed as TEXMFAUXDIRS is already included in TEXMF

xworld21 replied:
> Actually, I think that all of this can be removed as TEXMFAUXDIRS is already included in TEXMF

It turns out that my `TEXMFAUXTREES` addition is not doing anything for `luaotfload`, it's just being ignored! There is no other equivalent variable where to fit the extra folder as far as I know. I'll have to consult the resolver code to be sure.

xworld21 replied:
...so if I move `$TEXMFAUXTREES` in *front* of `$TEXMFHOME`, everything works out exactly as it should without `OSFONTDIR` (checked with multiple fonts and verified that `TEXMFHOME` is still picked up correctly). As soon as the other test PR is merged, I'll add a test. I won't try to figure out why the position is relevant.

xworld21 committed:
texlive.combine: only derivations under .packages

xworld21 committed:
texlive.buildTeXLivePackage: switch to fake multi-output derivations for TeX Live packages

xworld21 committed:
texlive.tlpdb.nix: detect presence of info pages

xworld21 committed:
texlive.buildTeXLivePackage: add setup hook

The setup hook makes most packages usable as build inputs and raises an
evaluation error for those that are not safe for use outside of
texlive.combine.

xworld21 committed:
texlive.bin.core: remove obsolete setup hook

xworld21 committed:
texlive.combine: add setup hook

The setup hook sets the variables already used in the wrappers and
propagates their PATH dependencies.
